[#1959] Improvement(script): Decouple script and config, move rss-env.sh to conf.

maobaolong opened the pull request:
<!--
1. Title: [#<issue>] <type>(<scope>): <subject>
   Examples:
     - "[#123] feat(operator): support xxx"
     - "[#233] fix: check null before access result in xxx"
     - "[MINOR] refactor: fix typo in variable name"
     - "[MINOR] docs: fix typo in README"
     - "[#255] test: fix flaky test NameOfTheTest"
   Reference: https://www.conventionalcommits.org/en/v1.0.0/
2. Contributor guidelines:
   https://github.com/apache/incubator-uniffle/blob/master/CONTRIBUTING.md
3. If the PR is unfinished, please mark this PR as draft.
-->

### What changes were proposed in this pull request?

rss-env is a config purpose file, it should be move to conf folder
We can store conf in our config server, but maintain scripts in git repository.

### Why are the changes needed?

Scripts within bin folder should be immutable, and configure driven.

Fix: #1959

### Does this PR introduce _any_ user-facing change?

rss-env.sh is moved into `conf` folder

### How was this patch tested?

```
UNIFFLE_COORDINATOR_JAVA_OPTS=" -Daa=bb"  COORDINATOR_CONF_FILE=`pwd`/conf/coordinator.conf  UNIFFLE_COORDINATOR_XMX_SIZE=3g bin/start-coordinator.sh

UNIFFLE_SHUFFLE_SERVER_XMX_SIZE=4g bin/start-shuffle-server.sh 

bin/start-dashboard.sh 

```

rickyma commented:
LGTM overall. Details need to be tested and reviewed more carefully.

jerqi commented:
You should modify this https://github.com/apache/incubator-uniffle/blob/master/deploy/kubernetes/operator/examples/configuration.yaml

maobaolong commented:
> You should modify this https://github.com/apache/incubator-uniffle/blob/master/deploy/kubernetes/operator/examples/configuration.yaml

Thanks you for your review, but I'm sorry that I cannot get what place should I modify after go through this yaml file?

maobaolong commented:
> LGTM overall. Details need to be tested and reviewed more carefully.

@rickyma Thanks for your suggestion, I've done some tests.

- I run coordinator\server\dashboard with `UNIFFLE_XXX_JAVA_OPTS` `UNIFFLE_XXX_XMX_SIZE`
- I run coordinator\server\dashboard without any other extra EnvVar

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -60,15 +60,16 @@ set -u
-          -Xmx${XMX_SIZE:-8g} \
-          -Xms${XMX_SIZE:-8g} \
+UNIFFLE_COORDINATOR_XMX_SIZE=${UNIFFLE_COORDINATOR_XMX_SIZE:-${XMX_SIZE:-"8g"}}
+UNIFFLE_COORDINATOR_DEFAULT_JVM_ARGS=${UNIFFLE_COORDINATOR_DEFAULT_JVM_ARGS:-" -server \
```
The default gc log jvm Args is defined by someone but not suitable for all use cases, it is verbose and I want to remove or add some variables for different case. So we can give a default gc log args, but we'd better not forbidden users to do customize configuration, otherwise, there could be many customized scripts.

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -60,15 +60,16 @@ set -u
-JVM_ARGS=" -server \
-          -Xmx${XMX_SIZE:-8g} \
-          -Xms${XMX_SIZE:-8g} \
+UNIFFLE_COORDINATOR_XMX_SIZE=${UNIFFLE_COORDINATOR_XMX_SIZE:-${XMX_SIZE:-"8g"}}
```
How to define different XMX_SIZE for server and coordinator dashboard on
same node?

maobaolong replied:
Yeah, we deploy `coordinator and server` in different node too in the production env, but I start all services of coordinator in our test env and local laptop in same node, this can make it easy and welcome to new contributor of uniffle.

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -91,24 +92,24 @@ GC_LOG_ARGS_NEW=" -XX:+IgnoreUnrecognizedVMOptions \
 version=$($RUNNER -version 2>&1 | awk -F[\".] '/version/ {print $2}')
 if [[ "$version" -lt "9" ]]; then
-  GC_ARGS=$GC_LOG_ARGS_LEGACY
+  JVM_GC_ARGS=${JVM_GC_ARGS:-$GC_LOG_ARGS_LEGACY}
```
We suggest to let user use g1gc but there are many different gc type and different customized JDKS developed by some companies

maobaolong replied:
@advancedxy Thanks for your suggestion, I guess I have applied to all start-xxx scripts already? 
- bin/start-coordinator.sh
- bin/start-dashboard.sh
- bin/start-shuffle-server.sh

maobaolong replied:
I see, sorry to mis-understand your suggestion before, now i get it and keep same behavior now.

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -26,9 +26,9 @@ load_rss_env
 cd "$RSS_HOME"
 
-COORDINATOR_CONF_FILE="${RSS_CONF_DIR}/coordinator.conf"
+COORDINATOR_CONF_FILE=${COORDINATOR_CONF_FILE:-"${RSS_CONF_DIR}/coordinator.conf"}
```
Yes, I need this variable to specify another temporary config file which copied from  the default one and do some modified base on this. I have no right to modified the default config file, but I want to update some config and restart the server to watch the status

maobaolong replied:
we can set RSS_CONF_DIR by inline env var like this
```
RSS_CONF_DIR=/tmp/rss_conf/ bin/start_shuffle_server.sh
```

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -26,9 +26,9 @@ load_rss_env
+COORDINATOR_CONF_FILE=${COORDINATOR_CONF_FILE:-"${RSS_CONF_DIR}/coordinator.conf"}
 JAR_DIR="${RSS_HOME}/jars"
-LOG_CONF_FILE="${RSS_CONF_DIR}/log4j2.xml"
+LOG_CONF_FILE=${LOG_CONF_FILE:-"${RSS_CONF_DIR}/log4j2.xml"}
```
yeah, different log4j files can let me do some modifications without modify the original log4j file, e.g.   set log level for some logger to show more log or hide some  verbose logger.

Sometimes we are not allowed or convenient to modify the original files which belongs to administrator.

maobaolong commented:
> configuration.yaml

@advancedxy Thanks for your suggestion, I move it by #1962 PTAL

maobaolong commented:
@advancedxy Thanks for your carefully review and suggestion, and I'm glad you are interested in this change.

But, I have some different idea, reply it inline after your comments.

jerqi commented:
> > You should modify this https://github.com/apache/incubator-uniffle/blob/master/deploy/kubernetes/operator/examples/configuration.yaml
> 
> @jerqi Thanks you for your review, but I'm sorry that I cannot get what place should I modify after go through this yaml file?

You should add rss-env.sh to this yaml file. This is the config yaml of Uniffle.

maobaolong commented:
@jerqi Thanks for your reply!

You mean I should add a `rss-env.sh` with default config?

Like this:

```
  rss.env: |-
set -o pipefail
set -o nounset   # exit the script if you try to use an uninitialised variable
set -o errexit   # exit the script if any statement returns a non-true return value

XMX_SIZE="10g" # Shuffle Server JVM XMX size

```

maobaolong commented:
@jerqi Thanks for your explain, it give me some information, base on this and ask for some help from our colleague around me, I know you want me to add rss-env.sh into the `configuration.yaml` which operator will generate config files from it.

I added it now, please help to take another look, thanks!

maobaolong commented:
@advancedxy Thanks for your review and detailed explanation, after think twice, I revert COORDINATOR_CONF_FILE and LOG_CONF_FILE to prevent user to customize conf name. 

Left a point need to be discussed further more.

I want to support config different jvm args for different process(coordinator/dashboard/server) but do not introduce any other new config file or config folder.

So, I add some prefix to distinguish the config for different server process.

maobaolong commented:
@advancedxy Thanks for your suggested name, I use short prefix instead of the long verbose one, PTAL.

maobaolong commented on conf/rss-env.sh:
```diff
@@ -23,8 +23,9 @@ set -o errexit   # exit the script if any statement returns a non-true return va
 # JAVA_HOME=<java_home_dir>
 # HADOOP_HOME=<hadoop_home_dir>
-XMX_SIZE="80g" # Shuffle Server JVM XMX size
+XMX_SIZE="8g" # Coordinator/ShuffleServer JVM XMX size
```
Because 80g is to large for Coordinator and ShuffleServer

maobaolong commented on bin/start-shuffle-server.sh:
```diff
@@ -26,18 +26,19 @@ load_rss_env
+SHUFFLE_SERVER_CONF_FILE=${SHUFFLE_SERVER_CONF_FILE:-"${RSS_CONF_DIR}/server.conf"}
 JAR_DIR="${RSS_HOME}/jars"
-LOG_CONF_FILE="${RSS_CONF_DIR}/log4j2.xml"
+LOG_CONF_FILE=${LOG_CONF_FILE:-"${RSS_CONF_DIR}/log4j2.xml"}
```
done

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -91,24 +92,24 @@ GC_LOG_ARGS_NEW=" -XX:+IgnoreUnrecognizedVMOptions \
 version=$($RUNNER -version 2>&1 | awk -F[\".] '/version/ {print $2}')
 if [[ "$version" -lt "9" ]]; then
-  GC_ARGS=$GC_LOG_ARGS_LEGACY
+  JVM_GC_ARGS=${JVM_GC_ARGS:-$GC_LOG_ARGS_LEGACY}
```
done

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -91,24 +92,24 @@ GC_LOG_ARGS_NEW=" -XX:+IgnoreUnrecognizedVMOptions \
 
-UNIFFLE_COORDINATOR_JAVA_OPTS=${UNIFFLE_COORDINATOR_JAVA_OPTS:-""}
-$RUNNER ${UNIFFLE_COORDINATOR_JAVA_OPTS} $ARGS $JVM_ARGS $GC_ARGS -cp $CLASSPATH $MAIN_CLASS --conf "$COORDINATOR_CONF_FILE" $@ &
+COORDINATOR_JAVA_OPTS=${COORDINATOR_JAVA_OPTS:-""}
```
> hmm. I think we cannot revert this option if it's already released and used by end users?

Don't worry about this, this is introduced by me just a few days, we rename it now is a good time.



> Also, it has similar meaning/semantic of `COORDINATOR_DEFAULT_JVM_ARGS, which I think we should avoid adding these two confusing ones?

Rename to `COORDINATOR_BASE_JVM_ARGS`

maobaolong commented on bin/start-dashboard.sh:
```diff
@@ -26,9 +26,9 @@ load_rss_env 0
+DASHBOARD_CONF_FILE=${DASHBOARD_CONF_FILE:-"${RSS_CONF_DIR}/dashboard.conf"}
 JAR_DIR="${RSS_HOME}/jars"
-LOG_CONF_FILE="${RSS_CONF_DIR}/log4j2.xml"
+LOG_CONF_FILE=${LOG_CONF_FILE:-"${RSS_CONF_DIR}/log4j2.xml"}
```
done

maobaolong commented on README.md:
```diff
@@ -155,7 +155,7 @@ If you have packaged tgz with hadoop jars, the env of `HADOOP_HOME` is needn't s
      JAVA_HOME=<java_home>
      HADOOP_HOME=<hadoop home>
-     XMX_SIZE="16g"
+     UNIFFLE_COORDINATOR_XMX_SIZE="16g"
```
done, added comment after this.

maobaolong commented on README.md:
```diff
@@ -206,7 +206,7 @@ Deploy Steps:
      JAVA_HOME=<java_home>
      HADOOP_HOME=<hadoop home>
-     XMX_SIZE="80g"
+     UNIFFLE_SHUFFLE_SERVER_XMX_SIZE="80g"
```
done

maobaolong commented:
@advancedxy Thanks for your review again, sorry for losing to revert the var name previous commit, and revert it and addressed your comments by new commit now. PTAL.

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -91,24 +93,24 @@ GC_LOG_ARGS_NEW=" -XX:+IgnoreUnrecognizedVMOptions \
 version=$($RUNNER -version 2>&1 | awk -F[\".] '/version/ {print $2}')
 if [[ "$version" -lt "9" ]]; then
-  GC_ARGS=$GC_LOG_ARGS_LEGACY
+  JVM_GC_ARGS="${JVM_GC_ARGS} ${GC_LOG_ARGS_LEGACY}"
```
This is great, i'd like to let user override jvm args as much as possible.

For this case, as we let user able to customize `JVM_GC_ARGS`, we have to add `PREFIX` to this.

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -91,24 +93,24 @@ GC_LOG_ARGS_NEW=" -XX:+IgnoreUnrecognizedVMOptions \
+  JVM_GC_ARGS="${JVM_GC_ARGS} ${GC_LOG_ARGS_LEGACY}"
 else
-  GC_ARGS=$GC_LOG_ARGS_NEW
+  JVM_GC_ARGS="${JVM_GC_ARGS} ${$GC_LOG_ARGS_NEW}"
```
Sorry for careless

maobaolong commented on bin/start-dashboard.sh:
```diff
@@ -72,24 +74,24 @@ GC_LOG_ARGS_NEW=" -XX:+IgnoreUnrecognizedVMOptions \
 if [[ "$version" -lt "9" ]]; then
-  GC_ARGS=$GC_LOG_ARGS_LEGACY
+  JVM_GC_ARGS="${JVM_GC_ARGS} ${GC_LOG_ARGS_LEGACY}"
 else
-  GC_ARGS=$GC_LOG_ARGS_NEW
+  JVM_GC_ARGS="${JVM_GC_ARGS} ${$GC_LOG_ARGS_NEW}"
```
done

maobaolong commented on deploy/kubernetes/docker/rss-env.sh:
```diff
@@ -40,4 +41,23 @@ JPS="${JAVA_HOME}/bin/jps"
 # RSS_LOG_DIR, Where log files are stored (Default: ${RSS_HOME}/logs)
+
+# Variables for the dashboard
+# DASHBOARD_JAVA_OPTS, JVM options for the dashboard, will override the default JVM options
```
> It might be too large for coordinator, but it should be normal for ShuffleServers to have max heap size of 80G.

We have to use the new VAR with PREFIX.

```
XMX_SIZE="8g" # Coordinator/Dashboard JVM XMX size
SHUFFLE_SERVER_XMX_SIZE="00g" # ShuffleServer JVM XMX size
```

maobaolong commented on bin/start-coordinator.sh:
```diff
@@ -91,24 +93,24 @@ GC_LOG_ARGS_NEW=" -XX:+IgnoreUnrecognizedVMOptions \
-UNIFFLE_COORDINATOR_JAVA_OPTS=${UNIFFLE_COORDINATOR_JAVA_OPTS:-""}
-$RUNNER ${UNIFFLE_COORDINATOR_JAVA_OPTS} $ARGS $JVM_ARGS $GC_ARGS -cp $CLASSPATH $MAIN_CLASS --conf "$COORDINATOR_CONF_FILE" $@ &
+COORDINATOR_JAVA_OPTS=${COORDINATOR_JAVA_OPTS:-""}
+$RUNNER ${JVM_LOG_ARGS} ${COORDINATOR_BASE_JVM_ARGS} ${JVM_GC_ARGS} ${COORDINATOR_JAVA_OPTS} -cp ${CLASSPATH} ${MAIN_CLASS} --conf "${COORDINATOR_CONF_FILE}" $@ &
```
OK, I put it after `COORDINATOR_JVM_GC_ARGS`

maobaolong commented on deploy/kubernetes/operator/examples/configuration.yaml:
```diff
@@ -93,4 +93,12 @@ data:
+    set -o errexit   # exit the script if any statement returns a non-true return value
+    HADOOP_HOME="/data/rssadmin/hadoop"
+    RUNNER="${JAVA_HOME}/bin/java"
+    JPS="${JAVA_HOME}/bin/jps"
```
Done

maobaolong commented:
@advancedxy Thank you for the carefully review for this PR! I addressed your comments and pushed changes, PTAL.

maobaolong commented:
@advancedxy Thanks for these previous review, and glad you approved it! 

I updated the title and description.

maobaolong commented:
@advancedxy Thank you advance for your careful and kind review!
